beacon-chain/cache/depositsnapshot/deposit_tree.go

```diff
@@ -0,0 +1,157 @@
+		mixInLength:             0,
+		finalizedExecutionBlock: executionBlock{},
+	}
+}
+
+// getSnapshot returns a deposit tree snapshot.
+//
+//nolint:unused
+func (d *DepositTree) getSnapshot() (DepositTreeSnapshot, error) {
+	if d.finalizedExecutionBlock == (executionBlock{}) {
+		return DepositTreeSnapshot{}, ErrEmptyExecutionBlock
+	}
+	var finalized [][32]byte
+	depositCount, _ := d.tree.GetFinalized(finalized)
+	return fromTreeParts(finalized, depositCount, d.finalizedExecutionBlock), nil
+}
+
+// fromSnapshot returns a deposit tree from a deposit tree snapshot.
+//
+//nolint:unused
+func fromSnapshot(snapshot DepositTreeSnapshot) (DepositTree, error) {
+	root, err := snapshot.CalculateRoot()
+	if err != nil {
+		return DepositTree{}, err
+	}
+	if snapshot.depositRoot != root {
+		return DepositTree{}, ErrInvalidSnapshotRoot
+	}
+	if snapshot.depositCount >= math.PowerOf2(uint64(DepositContractDepth)) {
+		return DepositTree{}, ErrTooManyDeposits
+	}
+	tree, err := fromSnapshotParts(snapshot.finalized, snapshot.depositCount, DepositContractDepth)
+	if err != nil {
+		return DepositTree{}, err
+	}
+	if snapshot.depositCount == 0 {
+		return DepositTree{}, ErrNoDeposits
+	}
+	return DepositTree{
+		tree:                    tree,
+		mixInLength:             snapshot.depositCount,
+		finalizedExecutionBlock: snapshot.executionBlock,
+	}, nil
+}
+
+// finalize marks a deposit as finalized.
+//
+//nolint:unused
+func (d *DepositTree) finalize(eth1data *eth.Eth1Data, executionBlockHeight uint64) error {
+	var blockHash [32]byte
+	copy(blockHash[:], eth1data.BlockHash)
+	d.finalizedExecutionBlock = executionBlock{
+		Hash:  blockHash,
+		Depth: executionBlockHeight,
+	}
+	_, err := d.tree.Finalize(eth1data.DepositCount, DepositContractDepth)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// getProof returns the Deposit tree proof.
+//
+//nolint:unused
+func (d *DepositTree) getProof(index uint64) ([32]byte, [][32]byte, error) {
+	if d.mixInLength <= 0 {
+		return [32]byte{}, nil, ErrInvalidMixInLength
+	}
+	finalizedDeposits, _ := d.tree.GetFinalized([][32]byte{})
+	if finalizedDeposits == 0 {
+		return [32]byte{}, nil, ErrNoFinalizedDeposits
+	}
+	if finalizedDeposits != 0 {
+		finalizedDeposits = finalizedDeposits - 1
+	}
+	if index <= finalizedDeposits {
+		return [32]byte{}, nil, ErrInvalidIndex
+	}
+	leaf, proof := generateProof(d.tree, index, DepositContractDepth)
+	var mixInLength [32]byte
+	copy(mixInLength[:], bytesutil.Uint64ToBytesLittleEndian32(d.mixInLength))
+	proof = append(proof, mixInLength)
+	return leaf, proof, nil
+}
+
+// getRoot returns the root of the deposit tree.
+//
+//nolint:unused
+func (d *DepositTree) getRoot() [32]byte {
+	root := d.tree.GetRoot()
+	return sha256.Sum256(append(root[:], bytesutil.Uint64ToBytesLittleEndian32(d.mixInLength)...))
+}
+
+// pushLeaf adds a new leaf to the tree.
+//
+//nolint:unused
+func (d *DepositTree) pushLeaf(leaf [32]byte) error {
+	var err error
+	d.mixInLength++
```

Comment:
should we add after we push? or do so inside of push leaf? worried that if it fails, we would have still modified the tree